### PR TITLE
Add GitLab ResourceClaims to provision Public Clud managed services

### PR DESCRIPTION
**Pull request by ichekrygin**
Dependent managed service resource claims:
- KubernetesCluster
- PostgreSQLInstance
- RedisCluster
- Buckets

### CODE REVIEW

File: pkg/controller/gitlab/add.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gitlab
+
+import (
+	xpcachev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/cache/v1alpha1"
+	xpcomputrev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/compute/v1alpha1"
```
Comment (negz) [inline pkg/controller/gitlab/add.go]: Typo in compute.

File: pkg/controller/gitlab/gitlab_test.go [hunk 2]
```diff
@@ -20,14 +20,18 @@ import (
...
 	"testing"
 
-	xpcorev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
-	"github.com/go-test/deep"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/google/go-cmp/cmp"
```
Comment (negz) [inline pkg/controller/gitlab/gitlab_test.go]: Nit: lots of whitespace around these imports.

File: pkg/controller/gitlab/gitlab.go [hunk 3]
```diff
@@ -20,67 +20,45 @@ import (
...
-}
-
-var _ reconcile.Reconciler = &Reconciler{}
+var log = logging.Logger.WithName("controller." + controllerName)
 
 // Reconciler reconciles a GitLab object
 type Reconciler struct {
 	client.Client
+	scheme *runtime.Scheme
+	mill   reconcilerMill
```
Comment (negz) [inline pkg/controller/gitlab/gitlab.go]: Is this a `factory` joke? 🤔

File: pkg/controller/gitlab/util.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	content, err := runtime.DefaultUnstructuredConverter.ToUnstructured(o)
+	if err != nil {
+		return nil, err
+	}
+	u := &unstructured.Unstructured{}
+	u.SetUnstructuredContent(content)
+	return u, nil
+}
+
+//// templates slice converted from the runtime.Object map to local cluster (namespace, bucketName) template objects
```
Comment (negz) [inline pkg/controller/gitlab/util.go]: Should we remove this commented code before merging?

File: pkg/controller/gitlab/util_test.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+*/
+
+package gitlab
+
+import (
+	"testing"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
```
Comment (negz) [inline pkg/controller/gitlab/util_test.go]: More import whitespace.

File: pkg/controller/gitlab/gitlab.go (hunk 3 above)
Comment (ichekrygin) [reply pkg/controller/gitlab/gitlab.go]: yes, it is.

File: design/initial-draft.md [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 ## Objective
 
-Provide functionality to support GitLab provisioning in hosted Crossplane environment by means of _GitLabController_ 
+Provide functionality to support GitLab provisioning in hosted Crossplane environment by means of _GitLabControler_
```
Comment (jbw976) [inline design/initial-draft.md]: typo: the old spelling of `_GitLabController_` was correct

File: design/initial-draft.md [hunk 7]
```diff
@@ -32,8 +32,8 @@ The initial workload implementation is at the “Proof of Concept” level.
 Currently, there is an active effort to redesign Workload to enable payload support for a much broader spectrum of 
 Kubernetes artifacts. At this time the working concepts replacing Workload are:
 
-*   _KubernetesApplication_
-*   _KubernetesApplicationResource_
+*   _KuberntesApplication_
```
Comment (jbw976) [inline design/initial-draft.md]: The old types are correct I think, and spelled correctly
```
_KubernetesApplication_
_KubernetesApplicationResource_
```
This change from KAR -> KAC happens in multiple places in this doc, is that intentional? in master i still see it as `KubernetesApplicationResource`, so this doc should be in agreement with whichever one is correct.

File: design/initial-draft.md [hunk 8]
```diff
@@ -137,13 +138,13 @@ GitLab application must produce portable workloads, i.e. the same workload must
 
 ### Environment
 
-GitLabController has set of expectation for the crossplane environment. These expectations must be satisfied before creating _GitLabApplication_ instance
+GitLabController has set of expectation for the crossplane environment. These expectations must be satisfied before creating _GitLabAapplication_ instance
```
Comment (jbw976) [inline design/initial-draft.md]: typo: the old spelling of `_GitLabApplication_` was correct

File: design/initial-draft.md [hunk 9]
```diff
@@ -155,7 +156,7 @@ While the _GitLabController_ Design currently being proposed it operates on term
 Helm installation is initially not portable, i.e. not cloud provider agnostic in the bucket credentials area:
 
 *   Connection secret
-*   s3cmd.config
+*   s3cmd.congig
```
Comment (jbw976) [inline design/initial-draft.md]: the old spelling of `config` was correct

File: pkg/controller/gitlab/kubernetes_test.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+				},
+				finder: &mockResourceClassFinder{
+					mockFind: func(ctx context.Context, provider v1.ObjectReference, resource string) (*v1.ObjectReference, error) {
+						return nil, nil
+					},
+				},
+			},
+			want: want{err: errors.Wrapf(testError, errorFmtFailedToCreate, kubernetesClaimKind, testKey)},
+		},
+		"FailToRetrievePostgresObject-Other": {
```
Comment (jbw976) [inline pkg/controller/gitlab/kubernetes_test.go]: this file is testing kubernetes resources, is the postgres reference here a copy/paste issue? i see it in some of the other test files too

**Comment by ichekrygin** [comment]
@jbw976 I must have a stale copy of the `initial-draft.md`, I have rebased and updated PTAL.

File: pkg/controller/gitlab/kubernetes_test.go (hunk 10 above)
Comment (ichekrygin) [reply pkg/controller/gitlab/kubernetes_test.go]: ah.. yes, must be. checking...

**Pull request merged**